src/google-classroom-extractor/google_classroom_extractor/mapping/user_submission_activities.py

```diff
@@ -0,0 +1,153 @@
+    history_df = submissions_df.explode(column="submissionHistory")
+
+    # expand submissionHistory dicts (stateHistory and gradeHistory) into their own columns
+    history_df = history_df["submissionHistory"].apply(Series).merge(history_df, left_index=True, right_index=True, how='outer')
+    history_df.drop(columns=["submissionHistory"], inplace=True)
+
+    # expand stateHistory (can assume exists, should always have at least one "CREATED" entry)
+    user_submission_df = concat([history_df, history_df["stateHistory"].apply(Series)], axis=1)
+    user_submission_df.dropna(subset=["stateHistory"], inplace=True)
+
+    # enrich stateHistory
+    user_submission_df["SourceSystemIdentifier"] = "S-" + user_submission_df[
+        ["courseId", "courseWorkId", "id", "stateTimestamp"]
+    ].agg("-".join, axis=1)
+
+    user_submission_df = user_submission_df[
+        [
+            "SourceSystemIdentifier",
+            "AssignmentIdentifier",
+            "stateTimestamp",
+            "state",
+            "courseId",
+            "actorUserId",
+        ]
+    ]
+
+    user_submission_df = user_submission_df.rename(
+        columns={
+            "stateTimestamp": "ActivityDateTime",
+            "state": "ActivityStatus",
+            "courseId": "LMSSectionIdentifier",
+            "actorUserId": "LMSUserIdentifier",
+        }
+    )
+
+    user_submission_df["ActivityType"] = ACTIVITY_TYPE_STATE
+
+    # expand gradeHistory if exists
+    if "gradeHistory" in history_df:
+        grade_history_df = concat([history_df, history_df["gradeHistory"].apply(Series)], axis=1)
+        grade_history_df.dropna(subset=["gradeHistory"], inplace=True)
+
+        # enrich gradeHistory
+        grade_history_df["SourceSystemIdentifier"] = "G-" + grade_history_df[
+            ["courseId", "courseWorkId", "id", "gradeTimestamp"]
+        ].agg("-".join, axis=1)
+
+        grade_history_df = grade_history_df[
+            [
+                "SourceSystemIdentifier",
+                "AssignmentIdentifier",
+                "gradeTimestamp",
+                "gradeChangeType",
+                "courseId",
+                "actorUserId",
+            ]
+        ]
+
+        grade_history_df = grade_history_df.rename(
+            columns={
+                "gradeTimestamp": "ActivityDateTime",
+                "gradeChangeType": "ActivityStatus",
+                "courseId": "LMSSectionIdentifier",
+                "actorUserId": "LMSUserIdentifier",
+            }
+        )
+
+        grade_history_df["ActivityType"] = ACTIVITY_TYPE_GRADE
+
+        # combine with stateHistory
+        user_submission_df = user_submission_df.append(grade_history_df)
+
+    # teacher actions can show up on student histories and vice-versa
+    user_submission_df.drop_duplicates(subset=["SourceSystemIdentifier"], inplace=True)
+
+    # finish with common columns
+    user_submission_df["LMSUserActivityIdentifier"] = user_submission_df["SourceSystemIdentifier"]
```

Comment:
What is the purpose of this duplicated column? I didn't catch in the initial review, only when looking at the actual output. I did not expect this column, `LMSUserActivityIdentifier`.

Side note: the sample user activities file was not kept up to date with actual UDM changes and thus has invalid columns. I can update it in my current ticket.

Reply:
Looking closely at the [UDM diagram](https://techdocs.ed-fi.org/display/EDFIODS/Fizz+-+LMS+Unified+Data+Model) again, I see that the `LMSUserActivity` table is described as having a primary key column `UserLMSActivityIdentifier`. Of course there's a small typo, as it should be `LMSUserActivityIdentifier` as you have it above. But I think this column would only show up in the RDBMS, as an auto-incrementing integer.

Reply:
I'll drop this.